optimism/src/main.rs

```diff
@@ -40,12 +40,113 @@ pub fn main() -> ExitCode {
 
     env_logger::Builder::from_env(env_logger::Env::default().default_filter_or("info")).init();
 
-    let mut env = witness::Env::<ark_bn254::Fq>::create(cannon::PAGE_SIZE as usize, state, po);
+    let domain_size = 1 << 15;
+
+    let domain =
+        kimchi::circuits::domains::EvaluationDomains::<ark_bn254::Fr>::create(domain_size).unwrap();
+
+    let srs = {
+        use ark_ff::UniformRand;
+
+        // Trusted setup toxic waste
+        let x = ark_bn254::Fr::rand(&mut rand::rngs::OsRng);
+
+        let mut srs = poly_commitment::pairing_proof::PairingSRS::create(x, domain_size);
+        srs.full_srs.add_lagrange_basis(domain.d1);
+        srs
+    };
+
+    let mut env = witness::Env::<ark_bn254::Fr>::create(cannon::PAGE_SIZE as usize, state, po);
+
+    let mut accumulator = proof::ProofInputs::<
+        ark_ec::short_weierstrass_jacobian::GroupAffine<ark_bn254::g1::Parameters>,
+    >::default();
+
+    let new_chunk = || proof::WitnessColumns {
+        scratch: std::array::from_fn(|_| Vec::with_capacity(domain_size)),
+        instruction_counter: Vec::with_capacity(domain_size),
+        error: Vec::with_capacity(domain_size),
+    };
+
+    let mut current_chunk = new_chunk();
+
+    use mina_poseidon::{
+        constants::PlonkSpongeConstantsKimchi,
+        sponge::{DefaultFqSponge, DefaultFrSponge},
+    };
+    type Fp = ark_bn254::Fr;
+    type SpongeParams = PlonkSpongeConstantsKimchi;
+    type BaseSponge = DefaultFqSponge<ark_bn254::g1::Parameters, SpongeParams>;
+    type ScalarSponge = DefaultFrSponge<Fp, SpongeParams>;
 
     while !env.halt {
         env.step(&configuration, &meta, &start);
+        for (scratch, scratch_chunk) in env
+            .scratch_state
+            .iter()
+            .zip(current_chunk.scratch.iter_mut())
+        {
+            scratch_chunk.push(*scratch);
+        }
+        current_chunk
+            .instruction_counter
+            .push(ark_bn254::Fr::from(env.instruction_counter));
+        // TODO
+        use ark_ff::UniformRand;
+        current_chunk
+            .error
+            .push(ark_bn254::Fr::rand(&mut rand::rngs::OsRng));
+        if current_chunk.instruction_counter.len() == 1 << 15 {
```

Comment:
Oh, or does chunk refer to each of the instances that will be folded?

Reply:
Like, in case we wanted to split the full circuit into instances of equal length that is a divisor of the circuit size?